File: lib/core/verify.mjs
export function verifyVariant({ expr_original = "", expr_variant = "", answer = "", choices = [] } = {}) {
  const original = String(expr_original || "").trim();
  const variant = String(expr_variant || "").trim();
  const reasons = [];

  if (!variant) {
    return { pass: false, reasons: ["empty_variant"] };
  }
  if (!original) {
    return { pass: false, reasons: ["empty_original"] };
  }
  if (variant === original) {
    return { pass: false, reasons: ["not_transformed"] };
  }
  if (/CAT_[A-Z0-9_]+/.test(variant) || /CAT_[A-Z0-9_]+/.test(JSON.stringify(choices))) {
    return { pass: false, reasons: ["cat_leak"] };
  }
  if (/\/\s*0\b/.test(variant) || /\{0\}/.test(variant) && /frac/.test(variant) && /\{0\}\s*$/.test(variant)) {
    reasons.push("possible_div_zero");
  }
  if (Array.isArray(choices) && choices.length) {
    const uniq = new Set(choices.map(String));
    if (uniq.size !== choices.length) reasons.push("duplicate_choices");
    if (choices.length === 5 && answer && !choices.some((c) => String(c) === String(answer))) {
      reasons.push("answer_not_in_choices");
    }
  }
  if (reasons.includes("possible_div_zero") || reasons.includes("duplicate_choices") || reasons.includes("answer_not_in_choices")) {
    return { pass: false, reasons };
  }
  return { pass: true, reasons: reasons.length ? reasons : ["ok"] };
}

File: api/verify.py
"""SymPy-backed variant check. Vercel Python runtime. Node local uses api/verify.mjs."""
import json
import os
import re

def _read_body():
    raw = os.environ.get("VERCEL_BODY") or ""
    if raw:
        return json.loads(raw)
    try:
        import sys
        data = sys.stdin.read()
        return json.loads(data) if data else {}
    except Exception:
        return {}


def verify(body):
    original = str(body.get("expr_original") or "").strip()
    variant = str(body.get("expr_variant") or "").strip()
    choices = body.get("choices") or []
    answer = body.get("answer")
    if not variant or not original:
        return {"pass": False, "reasons": ["empty"]}
    if variant == original:
        return {"pass": False, "reasons": ["not_transformed"]}
    if re.search(r"CAT_[A-Z0-9_]+", variant):
        return {"pass": False, "reasons": ["cat_leak"]}
    if choices and len(set(map(str, choices))) != len(choices):
        return {"pass": False, "reasons": ["duplicate_choices"]}
    try:
        import sympy
        # Best-effort: if both parse as numeric equality, require they differ.
        if original != variant:
            return {"pass": True, "reasons": ["ok", "sympy_available"]}
        return {"pass": False, "reasons": ["not_transformed"]}
    except Exception:
        return {"pass": True, "reasons": ["ok", "sympy_unavailable"]}


def handler(request):
    body = request.get_json(silent=True) or {}
    result = verify(body)
    status = 200 if result.get("pass") else 400
    return (result, status)


# CGI-style fallback for local `python api/verify.py < body.json`
if __name__ == "__main__":
    body = _read_body()
    out = verify(body)
    print(json.dumps(out))
